flaky test: `TestReorgUnsafeHead`

**Claude:** Filed by Claude on behalf of @ajsutton.

**CI run:** https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/123078/workflows/4327c7d8-ab80-44a4-9508-0c99d5bc2fe0/jobs/4868463/tests

**Job:** `memory-all-kona-op-reth` (acceptance tests)
**Date:** 2026-04-21
**Branch:** `gh-readonly-queue/develop/pr-20165-4df50ca0801bf0dd7049995d5a7c8cc8396023df` (merge queue)
**Test file:** `op-acceptance-tests/tests/interop/reorgs/unsafe_head_test.go:113`

**Error:**
```
Error Trace: /home/circleci/project/op-acceptance-tests/tests/interop/reorgs/unsafe_head_test.go:113
Error:       Should not be: 0xae78b3b90db8d7d10c044465707fc30e969695515bb3e778accb1554867a7831
Test:        TestReorgUnsafeHead
Messages:    Expected to get different heads on divergence block number, but got the same hash, so no reorg happened on chain A
```

Preceding log lines:

```
unsafe_head_test.go:111: Reorged chain A on divergence block number (prior the reorg) number=9 head=ae78b3..7a7831 parent=5aaa56..e2828e
unsafe_head_test.go:112: Reorged chain A on divergence block number (after the reorg) number=9 head=ae78b3..7a7831 parent=5aaa56..e2828e
```

**Context:**
- The test was unskipped and migrated to `TwoL2SupernodeInterop` in #20136 (commit `4df50ca080`).
- After `StopSequencer()`, the test uses `op-test-sequencer` to sequence a conflicting block at the unsafe head height, then calls `StartSequencer()` on the CL and waits for a new block.
- At the assertion, the block at the divergence height is still the original block — the EL's view of the unsafe head was never replaced by the test-sequencer's conflicting block (or the CL's resumed sequencing reverted the reorg).
- The test intersperses two `time.Sleep(2 * time.Second)` calls between `ia.New()` / `ia.Next()` at lines 97 and 101 — likely a timing race.

Will investigate and propose a fix.

**Claude:** Recurred on pipeline 125864 (workflow `c962d976-298b-48e2-b393-3b875bbfa9b5`, job `5087023` — `memory-all-opn-op-reth`, 2026-05-21).

CircleCI Insights shows 7 flakes for this test in the last window.

**This run's failure is a different symptom from the original issue.** The original was `Should not be: 0x...` (no reorg happened). This run is:

\`\`\`
Error Trace: /home/circleci/project/op-service/testreq/require.go:64
Error:       Condition never satisfied
Messages:    awaiting balance to be updated
--- FAIL: TestReorgUnsafeHead (50.42s)
\`\`\`

The faucet logs at the same wall-clock time show:
\`\`\`
faucet.go:154: ERROR failed to send funds err="context canceled"
faucet.go:45:  WARN  Failed to fund address ... err="Post .../faucet/dev-faucet-901: context deadline exceeded"
\`\`\`

So the test is failing in setup — `WaitForBalance` (op-devstack/dsl/eoa.go:228) never sees the faucet-funded balance arrive within `TransactionTimeout`. Same symptom is also affecting `TestReorgInitExecMsg` in the same job. Investigating as a fresh flake; will update this issue with root cause findings.

CI link: https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/125864/workflows/c962d976-298b-48e2-b393-3b875bbfa9b5/jobs/5087023/tests

**Claude:** Adding CI run context for this flake.

**CI run:** https://app.circleci.com/pipelines/github/ethereum-optimism/optimism/129017/workflows/003c8417-71e9-4f63-ad18-8060ac647964/jobs/5265092/tests
**Date:** 2026-06-26
**Error:**
```
assertions.go:387:
    Error Trace:	/home/circleci/project/op-devstack/dsl/l2_el.go:248
    Error:      	Received unexpected error:
        operation failed permanently after 30 attempts: expected head to reorg 0xbcd8ede64eb5859c91d564c81bef911be863509c5156cba695fd2d836ad564f4:6, but got 0xbcd8ede64eb5859c91d564c81bef911be863509c5156cba695fd2d836ad564f4:6
--- FAIL: TestReorgUnsafeHead (70.79s)
```

Note: the expected and actual head hashes are identical — the test waited for a reorg of the unsafe head that never happened (head stayed at `…564f4:6`), exhausting all 30 retry attempts.

Encountered while working on `klopfenstein/karst-estimategas-tx-gas-cap`.

--- a/op-node/rollup/sequencing/sequencer.go
+++ b/op-node/rollup/sequencing/sequencer.go
@@ -377,6 +377,10 @@ func (d *Sequencer) onPayloadSuccess(x engine.PayloadSuccessEvent) {
 
 func (d *Sequencer) onSequencerAction(ev SequencerActionEvent) {
 	d.log.Debug("Sequencer action")
+	if !d.active.Load() {
+		d.log.Debug("Ignoring stale sequencer action while inactive")
+		return
+	}
 	payload := d.asyncGossip.Get()
 	if payload != nil {
 		if d.latest.Info.ID == (eth.PayloadID{}) {

--- a/op-node/rollup/sequencing/sequencer_test.go
+++ b/op-node/rollup/sequencing/sequencer_test.go
@@ -260,6 +260,61 @@ func TestSequencer_StartStop(t *testing.T) {
 	require.NoError(t, err)
 }
 
+// TestSequencer_NoActionAfterStop verifies that a sequencer-action event that fires
+// after Stop() (because the driver's timer was already armed before we deactivated)
+// is ignored, rather than starting a new block-building job. Acting on that stale
+// event would issue a forkchoice update reasserting our head and could undo a reorg
+// introduced externally after we stopped. Regression test for issue #20198.
+func TestSequencer_NoActionAfterStop(t *testing.T) {
+	logger := testlog.Logger(t, log.LevelError)
+	seq, deps := createSequencer(logger)
+	emitter := &testutils.MockEmitter{}
+	seq.AttachEmitter(emitter)
+	deps.conductor.leader = true
+
+	testCtx := context.Background()
+	require.NoError(t, seq.Init(testCtx, false))
+	emitter.AssertExpectations(t)
+
+	// Bring latestSealed and latestHead to the same known block, so Stop() does not
+	// block waiting for the sealed block to catch up to the head.
+	head := eth.L2BlockRef{Hash: common.Hash{0xaa}}
+	envelope := &eth.ExecutionPayloadEnvelope{ExecutionPayload: &eth.ExecutionPayload{}}
+	emitter.ExpectOnce(engine.PayloadProcessEvent{Envelope: envelope, Ref: head})
+	seq.OnEvent(testCtx, engine.BuildSealedEvent{Envelope: envelope, Ref: head})
+	emitter.AssertExpectations(t)
+	seq.OnEvent(testCtx, engine.ForkchoiceUpdateEvent{UnsafeL2Head: head})
+	require.Equal(t, head.Hash, seq.latestHead.Hash)
+
+	// Start, then stop. While active the sequencer scheduled work; the driver may have
+	// already armed the sequencer-action timer at this point.
+	require.NoError(t, seq.Start(testCtx, head.Hash))
+	require.True(t, seq.Active())
+	_, ok := seq.NextAction()
+	require.True(t, ok, "active sequencer schedules work")
+
+	stopHead, err := seq.Stop(testCtx)
+	require.NoError(t, err)
+	require.Equal(t, head.Hash, stopHead)
+	require.False(t, seq.Active())
+
+	// Provide a valid L1 origin so that, absent the inactive-guard, onSequencerAction
+	// would proceed all the way to emitting a BuildStartEvent — turning a regression
+	// into a clear unexpected-emit failure rather than a panic.
+	deps.l1OriginSelector.l1OriginFn = func(l2Head eth.L2BlockRef) (eth.L1BlockRef, error) {
+		return eth.L1BlockRef{Number: 1000}, nil
+	}
+
+	// Fire the stale action. The MockEmitter has no pending expectations, so any emit
+	// here (i.e. the sequencer acting while stopped) fails the test.
+	seq.OnEvent(testCtx, SequencerActionEvent{})
+	emitter.AssertExpectations(t)
+
+	require.Equal(t, BuildingState{}, seq.latest, "no block-building job started while stopped")
+	_, ok = seq.NextAction()
+	require.False(t, ok, "stopped sequencer schedules no further work")
+}
+
 // TestSequencer_StaleBuild stops the sequencer after block-building,
 // but before processing the block locally,
 // and then continues it again, to check if the async-gossip gets cleared,

